safe_control_gym/safety_filters/cbf/cbf_nn.py

```diff
@@ -0,0 +1,359 @@
+
+        learned_residual = cs.dot(a_var.T, u_var) + b_var
+
+        right_hand_side = 0.0
+        slack_var = opti.variable(1, 1)
+
+        if self.soft_constrained:
+            # Quadratic objective
+            cost = 0.5 * cs.norm_2(uncertified_action_var - u_var) ** 2 + self.slack_weight * slack_var**2
+
+            # Soften CBF constraint
+            right_hand_side = slack_var
+
+            # Non-negativity constraint on slack variable
+            opti.subject_to(slack_var >= 0.0)
+        else:
+            opti.subject_to(slack_var == 0.0)
+            # Quadratic objective
+            cost = 0.5 * cs.norm_2(uncertified_action_var - u_var) ** 2
+
+        # CBF constraint
+        opti.subject_to(-self.linear_func(x=barrier_at_x)['y'] - lie_derivative_at_x - learned_residual <=
+                        right_hand_side)
+
+        # Input constraints
+        for input_constraint in self.input_constraints_sym:
+            opti.subject_to(input_constraint(u_var) <= 0)
+
+        opti.minimize(cost)
+
+        # Set verbosity option of optimizer
+        opts = {'printLevel': 'low', 'error_on_fail': False}
+
+        # Select QP solver
+        opti.solver('qpoases', opts)
+
+        self.opti_dict = {
+            'opti': opti,
+            'u_var': u_var,
+            'slack_var': slack_var,
+            'current_state_var': current_state_var,
+            'uncertified_action_var': uncertified_action_var,
+            'a_var': a_var,
+            'b_var': b_var,
+            'cost': cost
+        }
+
+    def solve_optimization(self,
+                           current_state,
+                           uncertified_action,
+                           ):
+        '''Solve the MPC optimization problem for a given observation and uncertified input.
+
+        Args:
+            current_state (ndarray): Current state/observation.
+            uncertified_action (ndarray): The uncertified_controller's action.
+
+        Returns:
+            certified_action (ndarray): The certified action.
+            feasible (bool): Whether the safety filtering was feasible or not.
+        '''
+
+        opti_dict = self.opti_dict
+        opti = opti_dict['opti']
+        u_var = opti_dict['u_var']
+        slack_var = opti_dict['slack_var']
+        current_state_var = opti_dict['current_state_var']
+        uncertified_action_var = opti_dict['uncertified_action_var']
+        a_var = opti_dict['a_var']
+        b_var = opti_dict['b_var']
+        cost = opti_dict['cost']
+
+        opti.set_value(current_state_var, current_state)
+        opti.set_value(uncertified_action_var, uncertified_action)
+
+        torch_state = torch.from_numpy(current_state)
+        torch_state = torch.unsqueeze(torch_state, 0)
+        torch_state = torch_state.to(torch.float32)
+        a_b = self.mlp(torch_state)
+        a_b = a_b.detach().numpy()
+        a = a_b[0, :self.model.nu]
+        b = a_b[0, -1]
```

Comment:
We could turn this into another method. We reuse this later as well

Reply:
Done.